### DISCUSSION TITLE
`getMappedType` behavior change in v6

### DISCUSSION THREAD

**Discussion by stefansundin** [mikro-orm/mikro-orm#5098] [General]
### Describe the bug

I'm not entirely sure if this is an actual bug or an intended change, but I couldn't find anything discussing it so I think it is worth opening an issue to resolve the question.

We have been using this configuration to automatically change `varchar` columns to `text`:

```typescript
  discovery: {
    getMappedType(type: string, platform: Platform) {
      // Change default string column type from `varchar` to `text`
      if (type === 'string') {
        return Type.getType(TextType);
      }
      return platform.getDefaultMappedType(type);
    },
  },
```

In mikro-orm 6.0.1, the `type` here will be `any` if the property is defined like this:

```typescript
  @Property({ nullable: true })
  company: string | null = null;
```

In 5.9.7 the `type` argument is `string`.

This seems to work though so perhaps I should update my entities to this syntax instead:

```typescript
  @Property({ nullable: true })
  company?: string;
```

Please tell me if this is what you'd prefer over the other snippet. Thanks!

Or maybe `getMappedType` should receive more information about the property so that an equivalent logic to the above can be written in v6.

### Reproduction

https://github.com/stefansundin/mikro-orm-reproduction

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.0.1

### Node.js version

v18.19.0, typescript@5.3.3

### Operating system

macOS 14.2.1

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
> In mikro-orm 6.0.1, the type here will be any if the property is defined like this:
>   company: string | null = null;

That expression cannot provide a type via reflect-metadata, as that always fail on union types. This is nothing new, it worked the same in v5, what changed is that we don't blindly narrow that to string during discovery, so we can infer the type better later on.

So you were just depending on a changed behavior, which was setting default type to string on all properties that failed to infer one via reflect-metadata. I dont consider that a bug, and I dont have any fix for it, as again, you were depending on a default which changed on purpose.

So in a way, if you want to have a fix, just treat `type === 'any'` the same way as `type === 'string'`, that's how it worked in v5.

**Reply by stefansundin** [reply]
Thanks for the explanation!
